File: src/leetcode/h42/TrappingRainWater.go
package h42

/**
 * LeetCode 42
 */
func trap(height []int) int {
	left := 0
	right := len(height) - 1
	leftMax := 0
	rightMax := 0
	trapped := 0

	for left < right {
		if height[left] > height[right] {
			bar := min(height[left], rightMax)
			if bar > height[right] {
				trapped += bar - height[right]
			}
			rightMax = max(height[right], rightMax)
			right--
		} else {
			bar := min(height[right], leftMax)
			if bar > height[left] {
				trapped += bar - height[left]
			}
			leftMax = max(height[left], leftMax)
			left++
		}
	}
	return trapped
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: src/leetcode/h42/TrappingRainWater.js
/**
 * LeetCode 42
 *
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
  let left = 0;
  let right = height.length - 1;
  let leftMax = 0;
  let rightMax = 0;
  let water = 0;

  while (left < right) {
    if (height[left] <= height[right]) {
      if (height[left] < leftMax) {
        water += leftMax - height[left];
      } else {
        leftMax = height[left];
      }
      left++;
    } else {
      if (height[right] < rightMax) {
        water += rightMax - height[right];
      } else {
        rightMax = height[right];
      }
      right--;
    }
  }
  return water;
};
